source/NVDAObjects/UIA/excel.py

```diff
@@ -0,0 +1,458 @@
+)
+from comtypes import GUID
+from scriptHandler import script
+import ui
+from . import UIA
+
+
+class ExcelCustomProperties:
+	""" UIA 'custom properties' specific to Excel.
+	Once registered, all subsequent registrations will return the same ID value.
+	This class should be used as a singleton via ExcelCustomProperties.get()
+	to prevent unnecessary work by repeatedly interacting with UIA.
+	"""
+	#: Singleton instance
+	_instance: "Optional[ExcelCustomProperties]" = None
+
+	@classmethod
+	def get(cls) -> "ExcelCustomProperties":
+		"""Get the singleton instance or initialise it.
+		"""
+		if cls._instance is None:
+			cls._instance = cls()
+		return cls._instance
+
+	def __init__(self):
+		self.cellFormula = CustomPropertyInfo(
+			guid=GUID("{E244641A-2785-41E9-A4A7-5BE5FE531507}"),
+			programmaticName="CellFormula",
+			uiaType=UIAutomationType.String,
+		)
+
+		self.cellNumberFormat = CustomPropertyInfo(
+			guid=GUID("{626CF4A0-A5AE-448B-A157-5EA4D1D057D7}"),
+			programmaticName="CellNumberFormat",
+			uiaType=UIAutomationType.String,
+		)
+
+		self.hasDataValidation = CustomPropertyInfo(
+			guid=GUID("{29F2E049-5DE9-4444-8338-6784C5D18ADF}"),
+			programmaticName="HasDataValidation",
+			uiaType=UIAutomationType.Bool,
+		)
+
+		self.hasDataValidationDropdown = CustomPropertyInfo(
+			guid=GUID("{1B93A5CD-0956-46ED-9BBF-016C1B9FD75F}"),
+			programmaticName="HasDataValidationDropdown",
+			uiaType=UIAutomationType.Bool,
+		)
+
+		self.dataValidationPrompt = CustomPropertyInfo(
+			guid=GUID("{7AAEE221-E14D-4DA4-83FE-842AAF06A9B7}"),
+			programmaticName="DataValidationPrompt",
+			uiaType=UIAutomationType.String,
+		)
+
+		self.hasConditionalFormatting = CustomPropertyInfo(
+			guid=GUID("{DFEF6BBD-7A50-41BD-971F-B5D741569A2B}"),
+			programmaticName="HasConditionalFormatting",
+			uiaType=UIAutomationType.Bool,
+		)
+
+		self.commentReplyCount = CustomPropertyInfo(
+			guid=GUID("{312F7536-259A-47C7-B192-AA16352522C4}"),
+			programmaticName="CommentReplyCount",
+			uiaType=UIAutomationType.Int,
+		)
+
+		self.areGridLinesVisible = CustomPropertyInfo(
+			guid=GUID("{4BB56516-F354-44CF-A5AA-96B52E968CFD}"),
+			programmaticName="AreGridlinesVisible",
+			uiaType=UIAutomationType.Bool,
+		)
+
+
+class ExcelObject(UIA):
+	"""Common base class for all Excel UIA objects
+	"""
+	_UIAExcelCustomProps = ExcelCustomProperties.get()
+
+
+class ExcelCell(ExcelObject):
+
+	# selecting cells causes duplicate focus events
+	shouldAllowDuplicateUIAFocusEvent = True
+
+	name = ""
+	role = controlTypes.ROLE_TABLECELL
+	rowHeaderText = None
+	columnHeaderText = None
+
+	#: Typing information for auto-property: _get_areGridlinesVisible
+	areGridlinesVisible: bool
+
+	def _get_areGridlinesVisible(self) -> bool:
+		parent = self.parent
+		# There will be at least one grid element between the cell and the sheet.
+		# There could be multiple as there might be a data table defined on the sheet.
+		while parent.role == controlTypes.ROLE_TABLE:
```

Comment:
Make sure that this loop covers `parent is None`. I know it is very unlikely, but better safe than sorry,